### PR TITLE
Make string_view operator<< use rdbuf directly

### PR DESCRIPTION
**Pull request by glenfe**

### CODE REVIEW

**Comment by mclow** [comment]
I don't think you can do this.  `operator<<(` needs to set the `badbit` in the stream (and possibly throw) on failure. The current code does this by delegating to `write`.

**Comment by glenfe** [comment]
The destructor of `detail::os_sv_holder` sets `badbit` in the stream conditionally (upon any kind of failure, exceptions or otherwise). Any exceptions thrown are propagated.

File: include/boost/utility/string_ref.hpp [hunk 1]
```diff
@@ -424,48 +425,63 @@ namespace boost {
...
-                os.write(str.data(), size);
-                if (os.good())
-                    detail::sr_insert_fill_chars(os, alignment_size);
+        class sr_os_holder {
+        public:
+            explicit sr_os_holder(std::basic_ostream<charT, traits>& os)
+                : os_(&os) { }
+            ~sr_os_holder() {
+                if (os_) {
+                    os_->setstate(std::basic_ostream<charT, traits>::badbit);
```
Comment (Lastique) [inline include/boost/utility/string_ref.hpp]: If this throws, it will call `terminate` since the destructor is `noexcept` by default.

File: include/boost/utility/string_ref.hpp [hunk 2]
```diff
@@ -424,48 +425,63 @@ namespace boost {
...
-            if (w <= size)
-                os.write(str.data(), size);
-            else
-                detail::sr_insert_aligned(os, str);
-            os.width(0);
+        typedef std::basic_ostream<charT, traits> stream;
+        detail::sr_os_holder<charT, traits> hold(os);
+        typename stream::sentry entry(os);
+        if (entry) {
+            std::size_t width = os.width();
```
Comment (Lastique) [inline include/boost/utility/string_ref.hpp]: `os.width()` is `streamsize`, which is signed, this will generate a warning. Same is below, where `sputn` is involved.

File: include/boost/utility/string_ref.hpp [hunk 3]
```diff
@@ -424,48 +425,63 @@ namespace boost {
...
-            charT fill_chars[chunk_size];
-            std::fill_n(fill_chars, static_cast< std::size_t >(chunk_size), os.fill());
-            for (; n >= chunk_size && os.good(); n -= chunk_size)
-                os.write(fill_chars, static_cast< std::size_t >(chunk_size));
-            if (n > 0 && os.good())
-                os.write(fill_chars, n);
+        inline bool sr_os_fill(std::basic_ostream<charT, traits>& os, std::size_t count)
+        {
+            for (charT c = os.fill(); count > 0; --count) {
+                if (traits::eq_int_type(os.rdbuf()->sputc(c), traits::eof())) {
```
[the remark this answers is not in the document]
Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: Addressed by preserving the 8 charT fill.

File: include/boost/utility/string_ref.hpp (hunk 1 above)
Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: Corrected noexcept specifier on destructor.

File: include/boost/utility/string_ref.hpp (hunk 2 above)
Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: Used streamsize.

File: include/boost/utility/string_ref.hpp [hunk 4]
```diff
@@ -424,48 +425,66 @@ namespace boost {
...
-                os.write(str.data(), size);
-            else
-                detail::sr_insert_aligned(os, str);
-            os.width(0);
+        typedef std::basic_ostream<charT, traits> stream;
+        detail::sr_os_holder<charT, traits> hold(os);
+        typename stream::sentry entry(os);
+        if (entry) {
+            std::streamsize width = os.width();
+            std::streamsize size = str.size();
```
Comment (Lastique) [inline include/boost/utility/string_ref.hpp]: `str.size()` is `std::size_t`, this line needed not to change. IMHO, the original code was better in that it explicitly casted `std::streamsize` to `std::size_t` and then proceeded with unsigned integers everywhere since width should not be negative anyway.

File: include/boost/utility/string_ref.hpp [hunk 5]
```diff
@@ -25,7 +26,7 @@
 #include <algorithm>
 #include <iterator>
 #include <string>
-#include <iosfwd>
+#include <ios>
```
Comment (Lastique) [inline include/boost/utility/string_ref.hpp]: Is this change needed? `iosfwd` is more lightweight and it should be enough when you're not doing IO. When you do, you've already included `ios` via `ostream` for your stream.

Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: For the use of the std::streamsize type (clang/libc++ required it).

Comment (Lastique) [reply include/boost/utility/string_ref.hpp]: In fact, `ios` doesn't necessarily forward-declare `basic_ostream`, so the change is not correct.

Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: `ios` does declare `std::streamsize`.

Comment (Lastique) [reply include/boost/utility/string_ref.hpp]: What if you don't mention `std::streamsize` in the code? Otherwise we'd have to include both headers.

Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: Sounds good to me.

File: include/boost/utility/string_ref.hpp (hunk 4 above)
Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: Updated. Using `std::size_t` also avoids requiring `ios` in favor of `iosfwd`.

File: include/boost/utility/string_ref.hpp (hunk 2 above)
Comment (Lastique) [reply include/boost/utility/string_ref.hpp]: This still needs an explicit cast to silence the warning, see the original code. I think, it was an MSVC warning, but I can't remember the version.

Comment (glenfe) [reply include/boost/utility/string_ref.hpp]: Done (and thanks).
